mono/metadata/threads.c

```diff
@@ -1650,40 +1655,60 @@ mono_thread_set_name_internal (MonoInternalThread *this_obj, MonoString *name, g
 		mono_error_set_invalid_operation (error, "Thread.Name can only be set once.");
 		return;
 	}
-	if (this_obj->name) {
-		g_free (this_obj->name);
-		this_obj->name_len = 0;
-	}
-	if (name) {
-		this_obj->name = g_memdup (mono_string_chars (name), mono_string_length (name) * sizeof (gunichar2));
-		this_obj->name_len = mono_string_length (name);
 
+	g_free (this_obj->name);
+	this_obj->name = 0;
+	this_obj->name_len = 0;
+
+	if (length > 0) {
+		this_obj->name = utf16 ? (gunichar2*)g_memdup (utf16, length * sizeof (gunichar2))
+							   : g_utf8_to_utf16 (utf8, length, NULL, NULL, NULL);
+		this_obj->name_len = length;
 		if (permanent)
 			this_obj->flags |= MONO_THREAD_FLAG_NAME_SET;
 	}
-	else
-		this_obj->name = NULL;
 
 	if (!(this_obj->state & ThreadState_Stopped))
 		tid = thread_get_tid (this_obj);
 
 	UNLOCK_THREAD (this_obj);
 
 	if (this_obj->name && tid) {
-		char *tname = mono_string_to_utf8_checked (name, error);
-		return_if_nok (error);
+		const char *tname = 0;
+		if (utf8)
+			tname = utf8;
+		else {
+			tname = mono_utf16_to_utf8 (utf16, length, error);
+			return_if_nok (error);
+		}
 		MONO_PROFILER_RAISE (thread_name, ((uintptr_t)tid, tname));
 		mono_native_thread_set_name (tid, tname);
-		mono_free (tname);
+		if (!utf8)
+			mono_free (MONO_CONST_CAST(char*)(tname));
 	}
 }
 
+void
+mono_thread_set_name_cstr_internal (MonoInternalThread *thread, const char* name, gboolean permanent, gboolean reset, MonoError *error)
+{
+	mono_thread_set_name_utf_internal (thread, strlen(name), name, NULL, permanent, reset, error);
+}
+
 void 
-ves_icall_System_Threading_Thread_SetName_internal (MonoInternalThread *this_obj, MonoString *name)
+ves_icall_System_Threading_Thread_SetName_internal (
+	MonoInternalThreadHandle thread_handle, MonoStringHandle name_handle, MonoError *error)
 {
-	MonoError error;
-	mono_thread_set_name_internal (this_obj, name, TRUE, FALSE, &error);
-	mono_error_set_pending_exception (&error);
+	error_init (error);
+
+	g_assert (!MONO_HANDLE_IS_NULL(thread_handle));
```

Comment:
Could use `mono_error_set_argument_null(error, ...)` and to throw a NPE in managed code instead of asserting.

Reply:
Yes this was bothering me. Let me get back to it.. "ANE"? ArgumentNullException?
Is it particularly possible though for "this" pointer? I guess, at least, with reflection? And maybe in normal code. I'll do it shortly.

Reply:
ok, how about ?https://github.com/jaykrell/mono/commit/d815c874cd35c0cc369d8fd795bca6ba076d8064

Reply:
1. I'd avoid `MONO_CHECK_ARG_NULL`  in new code - it uses `MonoException*` objects and `mono_set_pending_exception` 
2. On the managed side (https://github.com/mono/mono/blob/381d11c97eb5fbeca9afbbefc518f79dc0eb8c4f/mcs/class/corlib/System.Threading/Thread.cs#L387-L403) it's actually not a `this` pointer. Unlikely to be NULL, but not otherwise null checked anywhere.